docs/design/features/byreflike-generics.md

```diff
@@ -43,18 +43,25 @@ namespace System.Runtime.CompilerServices
 }
 ```
 
-The compiler will need an indication for existing troublesome APIs where ByRefLike types will be permissable, but where the failure will be handled at runtime. An attribute will be created and added to these APIs.
+The compiler will need an indication for existing troublesome APIs where ByRefLike types will be permissable, but where the failure will be handled by the runtime. An attribute will be created and added to these APIs. The compiler will only respect the attribute that is defined in the same assembly containing `System.Object`.
 
 ```csharp
 namespace System.Runtime.CompilerServices
 {
     /// <summary>
-    /// Indicates to the compiler that constraint checks should be suppressed
-    /// and will instead be enforced at run-time.
+    /// Indicates to the compiler the ByRefLike constraint check should be suppressed.
     /// </summary>
-    [AttributeUsage(AttributeTargets.Constructor | AttributeTargets.Method | AttributeTargets.Property)]
-    internal sealed class SuppressConstraintChecksAttribute : Attribute
-    { }
+    /// <remarks>
+    /// The checking will be suppressed for both the signature and method body. These
+    /// checks are deferred and will be enforced at run-time.
+    /// </remarks>
+    /// <seealso href="https://github.com/dotnet/runtime/issues/99788">Design discussion</seealso>
+    [AttributeUsage(AttributeTargets.Constructor | AttributeTargets.Method | AttributeTargets.Property, Inherited = false, AllowMultiple = false)]
+    internal sealed class SuppressByRefLikeConstraintChecksAttribute : Attribute
+    {
+        /// <summary>Initializes the attribute.</summary>
+        public SuppressByRefLikeConstraintChecksAttribute() { }
+    }
 }
```

Comment:
Some reflection APIs are going to throw when inspecting these methods. I do not see a way around it. It may break some frequently used patterns like: `foreach overload { check whether the types match }`. We should at least make sure that the reflection internals are robust against this. 

For example, we should make sure that `typeof(Span<Span<int>>).GetConstructor(BindingFlags.Instance | BindingFlags.Public, [typeof(void*), typeof(int)])` does not leak the exception that may be throw by attempt to instantiate `Span<int>[]` used by other constructor overloads.

This does not need to be addressed in this PR. Just something to note later.

Reply:
Type in IL:

```
.class public sequential ansi sealed beforefieldinit InvalidCSharp.GS`1<byreflike T>
    extends [System.Runtime]System.ValueType
{
    .method public hidebysig specialname rtspecialname 
        instance void .ctor (
            void* p,
            int32 l
        ) cil managed 
    {
        ret;
    }

    .method public hidebysig specialname rtspecialname 
        instance void .ctor (
            !T[] a
        ) cil managed 
    {
        ret;
    }
}
```

```csharp
// ldtoken valuetype InvalidCSharp.GS`1<valuetype InvalidCSharp.GS`1<!!T>>
// call class [System.Runtime]System.Type [System.Runtime]System.Type::GetTypeFromHandle(valuetype [System.Runtime]System.RuntimeTypeHandle)
Type t = InvalidCSharp.Exec.GetInvalidTType<int>();
Console.WriteLine(t.GetConstructor(BindingFlags.Instance | BindingFlags.Public, [typeof(void*), typeof(int)]));

// Output:
// Void .ctor(Void*, Int32)
```

Reply:
Does it still work when you flip the order of the methods in IL? Does it work on all runtimes?

Reply:
It works regardless of order. I didn't try all the runtimes.